### PR TITLE
Datastore: Add support for 'TransactionOptions'

### PR DESCRIPTION
**Pull request by chemelnucfin**
Changes reflecting review and Luke's autogen (please see #4348 for issues regarding his autogen).  A manual fix has been included.

Closes #4335.  This includes reviews and changes from there.

### CODE REVIEW

**Comment by chemelnucfin** [comment]
hold on a second.  I am still missing a couple comments.

File: datastore/google/cloud/datastore_v1/gapic/datastore_client.py [hunk 1]
```diff
@@ -151,8 +151,8 @@ def __init__(self,
     # Service calls
     def lookup(self,
                project_id,
-               keys,
                read_options=None,
+               keys=None,
```
Comment (dhermes) [inline datastore/google/cloud/datastore_v1/gapic/datastore_client.py]: @chemelnucfin Is this change not in `master` yet? Wasn't that required for #4348?

File: datastore/google/cloud/datastore/transaction.py [hunk 2]
```diff
@@ -152,13 +153,19 @@ def Entity(*args, **kwargs):
...
     :param client: the client used to connect to datastore.
+
+    :type options: `google.cloud.datastore.transaction.TransactionOptions`
+    :param options: the transaction options used for this transaction.
     """
 
     _status = None
 
-    def __init__(self, client):
+    def __init__(self, client, options=None):
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: @chemelnucfin What became of our discussion about just using a `bool` here for `read_only`?

File: datastore/tests/system/test_system.py [hunk 3]
```diff
@@ -442,6 +443,25 @@ def test_transaction_via_with_statement(self):
         self.case_entities_to_delete.append(retrieved_entity)
         self.assertEqual(retrieved_entity, entity)
 
+    def test_readonly_put(self):
```
Comment (dhermes) [inline datastore/tests/system/test_system.py]: @chemelnucfin This **should not** be a system test. It is testing behavior of the client that **does not** depend on a live network.

File: datastore/tests/system/test_system.py [hunk 4]
```diff
@@ -442,6 +443,25 @@ def test_transaction_via_with_statement(self):
...
+        options = types.TransactionOptions(
+                            read_only=types.TransactionOptions.ReadOnly())
+        with self.assertRaises(RuntimeError):
+            with Config.CLIENT.transaction(options) as xact:
+                xact.put(entity)
+
+    def test_readwrite_put(self):
+        entity = datastore.Entity(key=Config.CLIENT.key('Company', 'Google'))
+        options = types.TransactionOptions(
+                            read_write=types.TransactionOptions.ReadWrite())
```
Comment (dhermes) [inline datastore/tests/system/test_system.py]: Please indent 4 spaces (not sure why linter didn't complain, sorry).

File: datastore/google/cloud/datastore/transaction.py [hunk 5]
```diff
@@ -152,13 +153,19 @@ def Entity(*args, **kwargs):
 
     :type client: :class:`google.cloud.datastore.client.Client`
     :param client: the client used to connect to datastore.
+
+    :type options: `google.cloud.datastore.transaction.TransactionOptions`
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: @chemelnucfin 

1. This class doesn't seem to exist?
1. In general we don't ask users to explicitly create protobuf objects, so you'd either need a wrapper class of your own, or you should do as we discussed before (just use `read_only=False` as the argument and create `TransactionOptions` protobuf for the user).
1. You should use `:class:` before a Sphinx type

File: datastore/google/cloud/datastore/transaction.py [hunk 6]
```diff
@@ -152,13 +153,19 @@ def Entity(*args, **kwargs):
...
     """
 
     _status = None
 
-    def __init__(self, client):
+    def __init__(self, client, options=None):
         super(Transaction, self).__init__(client)
         self._id = None
+        if options is None:
+            options = TransactionOptions()
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: `None` is an acceptable default for optional protobuf arguments (e.g. in GAPIC methods)

File: datastore/tests/unit/test_transaction.py [hunk 7]
```diff
@@ -12,22 +12,55 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import mock
 import unittest
 
-import mock
+
+class TestTransactionOptions(unittest.TestCase):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This test corresponds to an auto-generated class. It should already be tested.

File: datastore/tests/unit/test_transaction.py [hunk 8]
```diff
@@ -12,22 +12,55 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import mock
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This is not in the standard library, it should be in another imports section:

```python
import unittest

import mock
```

FWIW `unittest.mock` **is** in the standard library, but only in Python 3.

File: datastore/tests/unit/test_transaction.py [hunk 9]
```diff
@@ -12,22 +12,55 @@
...
+    def test_read_write(self):
+        options = self._make_one(read_write=
+                                 self._get_target_class().ReadWrite())
+        self.assertTrue(options.HasField('read_write'))
+
+    def test_previous_transaction(self):
+        previous_transaction = b'234'
+        options = self._make_one(
+                           read_write=self._get_target_class().ReadWrite(
+                               previous_transaction=previous_transaction))
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This method should be deleted? (Per my previous comment, this whole class should.)

----

Just for the sake of explanation: What rules is your editor using to make these indents? They don't seem to make sense? Please indent 4 spaces. I would **much** prefer to see:

```python
klass = self._get_target_class()
options = self._make_one(
    read_write=klass.ReadWrite(
        previous_transaction=previous_transaction,
    ),
)
```

Note also the call to `_get_target_class()` **before** it is used (there is such a thing as doing "too much" in a single expression).

File: datastore/tests/unit/test_transaction.py [hunk 10]
```diff
@@ -12,22 +12,55 @@
...
+        self.assertEqual(options.read_write.previous_transaction,
+                         previous_transaction)
 
 
 class TestTransaction(unittest.TestCase):
 
     @staticmethod
     def _get_target_class():
         from google.cloud.datastore.transaction import Transaction
-
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: Please don't make cosmetic changes like this unrelated to the PR

File: datastore/tests/unit/test_transaction.py [hunk 11]
```diff
@@ -212,6 +245,95 @@ class Foo(Exception):
...
+        entity = _Entity()
+        xact = self._make_one(client, options=options)
+        xact.begin()
+        xact.put(entity)
+        mutated_entity = _mutated_pb(self, xact.mutations, 'insert')
+        self.assertEqual(mutated_entity.key, entity.key.to_protobuf())
+
+
+def _assert_num_mutations(test_case, mutation_pb_list, num_mutations):
+    test_case.assertEqual(len(mutation_pb_list), num_mutations)
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This is a one-liner, ISTM you should just inline this logic wherever `_assert_num_mutations` is used?

File: datastore/tests/unit/test_transaction.py [hunk 12]
```diff
@@ -212,6 +245,95 @@ class Foo(Exception):
...
+            xact.put(entity)
+
+    def test_put_read_write(self):
+        project = 'PROJECT'
+        id_ = 943243
+        ds_api = _make_datastore_api(xact_id=id_)
+        client = _Client(project, datastore_api=ds_api)
+        options = self._make_options(read_write=
+                                     self._make_options().ReadWrite(
+                                        previous_transaction=b'321'))
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: @chemelnucfin Ditto here on the indents. Also **never** indent after an `=` sign.

File: datastore/tests/unit/test_transaction.py [hunk 13]
```diff
@@ -212,6 +245,95 @@ class Foo(Exception):
...
+        with self.assertRaises(RuntimeError):
+            xact.put(entity)
+
+    def test_put_read_write(self):
+        project = 'PROJECT'
+        id_ = 943243
+        ds_api = _make_datastore_api(xact_id=id_)
+        client = _Client(project, datastore_api=ds_api)
+        options = self._make_options(read_write=
+                                     self._make_options().ReadWrite(
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: ISTM you want `_get_options_class()` here, not `_make_options()`.

File: datastore/tests/unit/test_transaction.py [hunk 14]
```diff
@@ -212,6 +245,95 @@ class Foo(Exception):
...
+        begin = ds_api.begin_transaction
+        begin.assert_called_once()
+        self.assertEqual(begin.call_count, 1)
+
+    def test_put_read_only(self):
+        project = 'PROJECT'
+        id_ = 943243
+        ds_api = _make_datastore_api(xact_id=id_)
+        client = _Client(project, datastore_api=ds_api)
+        options = self._make_options(read_only=self._make_options().ReadOnly())
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: ISTM you want `_get_options_class()` here, not `_make_options()`.

File: datastore/tests/unit/test_transaction.py [hunk 15]
```diff
@@ -281,7 +403,6 @@ def _make_commit_response(*keys):
 def _make_datastore_api(*keys, **kwargs):
     commit_method = mock.Mock(
         return_value=_make_commit_response(*keys), spec=[])
-
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: Ditto here on cosmetic changes

File: datastore/google/cloud/datastore/transaction.py (hunk 2 above)
Comment (chemelnucfin) [reply datastore/google/cloud/datastore/transaction.py]: @lukesneeringer @dhermes So I just want to make sure what our consensus on this is.  Do we actually allow users to specify the entire TransactionOptions just like what it is in protobuf, or do we just make it easier for the users by just specifying a readonly boolean or something similar?  I originally thought that it would be best to follow the protobuf, but @dhermes mentioned it would be easier for the users if it was simpler.  So now I think I can be convinced either way.

I will shelve this until we reach a consensus on this.

Comment (tseaver) [reply datastore/google/cloud/datastore/transaction.py]: As a user, I would prefer to pass "simple" arguments, rather than having to construct a "complicated" options object.  E.g.:

```python
with client.transaction(read_only=True):
    ...
```

is much nicer than:

```python
from google.cloud.datastore import TransactionOptions
txn_options = TransactionOptions(read_only=True)
with client.transaction(options=txn_options):
    ...
```

Likewise for passing in the prior transaction ID (note that the two cases are exclusive).

The tradeoff comes when the number of things to be configured gets larger, or when they interact in subtle ways.

Comment (dhermes) [reply datastore/google/cloud/datastore/transaction.py]: @tseaver IMO we "shouldn't" even expose the `previous_transaction` field to users. It should be something we use internally when doing retries **for** users.

Comment (chemelnucfin) [reply datastore/google/cloud/datastore/transaction.py]: @lukesneeringer Could you chime in on this issue?  Especially related to (future) upstream developments or YAGNI.

Also regarding the previous transaction (RunInTransaction)

Comment (chemelnucfin) [reply datastore/google/cloud/datastore/transaction.py]: @lukesneeringer @dhermes @tseaver 
I thought about this issue some more.  I think I will continue to support that TransactionOptions is a standalone object as opposed to a boolean.  My reasoning is based on the fact that the signature will be hard to change after it sets in stone.  If it was a boolean, it would be hard to change that should TransactionOptions change in the future, which I think is in the plans or they would have used a boolean also for the readonly.

Comment (dhermes) [reply datastore/google/cloud/datastore/transaction.py]: So I just discussed with @schmidt-sebastian and have concluded that `TransactionOptions` would be preparing for a future that will never come. Please use a boolean `read_only` keyword argument.

File: datastore/tests/system/test_system.py (hunk 3 above)
Comment (chemelnucfin) [reply datastore/tests/system/test_system.py]: @dhermes how is this different than the test above it?

Comment (dhermes) [reply datastore/tests/system/test_system.py]: The test above it (`test_transaction_via_with_statement`) actually talks to the network.

**Comment by chemelnucfin** [comment]
I mean, wouldn't it be good to have one here to make sure it doesn't reach
the network (like it is doing what it should do)?

**Comment by dhermes** [comment]
No, we can test that **code path** with a unit test.

**Comment by chemelnucfin** [comment]
@dhermes @tseaver PTAL.

File: datastore/google/cloud/datastore/client.py [hunk 16]
```diff
@@ -134,7 +134,7 @@ def _extended_lookup(datastore_api, project, key_pbs,
         loop_num += 1
         lookup_response = datastore_api.lookup(
             project,
-            key_pbs,
+            keys=key_pbs,
```
Comment (dhermes) [inline datastore/google/cloud/datastore/client.py]: Please don't modify this. I intentionally changed it so that positional arguments would be passed as positional and keyword as keyword. (Also this is unrelated to your PR.)

File: datastore/google/cloud/datastore/client.py [hunk 17]
```diff
@@ -511,6 +511,10 @@ def batch(self):
         """Proxy to :class:`google.cloud.datastore.batch.Batch`."""
         return Batch(self)
 
+    def read_only_transaction(self):
```
Comment (dhermes) [inline datastore/google/cloud/datastore/client.py]: Please remove this factory and add `**kwargs` to the `Client.transaction()` factory. In addition, you might update that docstring to just say the same arguments accepted by `Transaction()` are accepted as keywords.

File: datastore/google/cloud/datastore/transaction.py [hunk 18]
```diff
@@ -152,13 +153,22 @@ def Entity(*args, **kwargs):
...
-    def __init__(self, client):
+    def __init__(self, client, read_only=False):
         super(Transaction, self).__init__(client)
         self._id = None
+        if not read_only:
+            options = TransactionOptions()
+        else:
+            options = TransactionOptions(
+                read_only=TransactionOptions.ReadOnly())
+        self.options = options
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: Please make this non-public (i.e. `self._options`)

File: datastore/google/cloud/datastore/transaction.py [hunk 19]
```diff
@@ -152,13 +153,22 @@ def Entity(*args, **kwargs):
...
+    :param read_only: indicates the transaction is read only.
     """
 
     _status = None
 
-    def __init__(self, client):
+    def __init__(self, client, read_only=False):
         super(Transaction, self).__init__(client)
         self._id = None
+        if not read_only:
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: This is definitely a nit-pick, but I would prefer you switch the order of the branches and drop the negation (it is unnecessary "work" to negate the boolean).

File: datastore/google/cloud/datastore/transaction.py [hunk 20]
```diff
@@ -234,3 +244,20 @@ def commit(self):
         finally:
             # Clear our own ID in case this gets accidentally reused.
             self._id = None
+
+    def put(self, entity):
+        """Ensures the transaction is not marked readonly and calls Batch.put
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: 1. This should not be primary description of what `put()` does. I would suggest "Add an entity to be committed." as the one-liner and then move this "Ensures ..." bit into the long description.
1. "readonly" -> "read only"

File: datastore/google/cloud/datastore/transaction.py [hunk 21]
```diff
@@ -234,3 +244,20 @@ def commit(self):
...
+
+        This method calls :meth:`put` in the superclass, `Batch`, as long as
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: This is user-facing documentation, so it's "inappropriate" to be discussing implementation details.

File: datastore/google/cloud/datastore/transaction.py [hunk 22]
```diff
@@ -234,3 +244,20 @@ def commit(self):
...
+        the transaction is not marked read only.
+
+        :type entity: :class:`google.cloud.datastore.entity.Entity`
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: I would suggest using a `~` before `google.cloud.datastore.entity.Entity` (it will make Sphinx only print `Entity` in rendered documentation)

File: datastore/google/cloud/datastore/transaction.py [hunk 23]
```diff
@@ -234,3 +244,20 @@ def commit(self):
...
+        :param entity: the entity to be saved.
+
+        :raises: :class:`RuntimeError` if the transaction
+                 is marked ReadOnly
+        """
+        if not self.options.HasField('read_only'):
```
Comment (dhermes) [inline datastore/google/cloud/datastore/transaction.py]: Ditto here on my "don't use `not`" nitpick

File: datastore/google/cloud/datastore_v1/gapic/datastore_client.py (hunk 1 above)
Comment (dhermes) [reply datastore/google/cloud/datastore_v1/gapic/datastore_client.py]: @chemelnucfin Why is this change here? The autogen was updated in #4348 so you should not be touching these files.

File: datastore/google/cloud/datastore_v1/gapic/datastore_client.py [hunk 24]
```diff
@@ -194,7 +194,7 @@ def lookup(self,
             ValueError: If the parameters are invalid.
         """
         request = datastore_pb2.LookupRequest(
-            project_id=project_id, keys=keys, read_options=read_options)
+            project_id=project_id, read_options=read_options, keys=keys)
```
Comment (dhermes) [inline datastore/google/cloud/datastore_v1/gapic/datastore_client.py]: Ditto here. Please remove edits from autogenerated files.

File: datastore/tests/system/test_system.py [hunk 25]
```diff
@@ -442,6 +443,17 @@ def test_transaction_via_with_statement(self):
         self.case_entities_to_delete.append(retrieved_entity)
         self.assertEqual(retrieved_entity, entity)
 
+    def test_readwrite_put(self):
```
Comment (dhermes) [inline datastore/tests/system/test_system.py]: IMO this is an unnecessary test. Transactions are read-write by default and we already have several tests which exercise transactions.

File: datastore/tests/system/test_system.py [hunk 26]
```diff
@@ -442,6 +443,17 @@ def test_transaction_via_with_statement(self):
...
+        entity = datastore.Entity(key=Config.CLIENT.key('Company', 'Google'))
+        options = types.TransactionOptions(
+            read_write=types.TransactionOptions.ReadWrite())
+        with Config.CLIENT.transaction() as xact:
+            xact.put(entity)
```
Comment (dhermes) [inline datastore/tests/system/test_system.py]: You never actually use the `options` variable (PyLint probably complains but maybe `flake8` doesn't).

If you wanted to actually use it you'd need to so `xact.options = options` (after making it non-public `xact._options = options`)

File: datastore/tests/unit/test_client.py [hunk 27]
```diff
@@ -353,7 +353,7 @@ def test_get_multi_miss(self):
         read_options = datastore_pb2.ReadOptions()
         ds_api.lookup.assert_called_once_with(
             self.PROJECT,
-            [key.to_protobuf()],
+            keys=[key.to_protobuf()],
```
Comment (dhermes) [inline datastore/tests/unit/test_client.py]: Ditto here (and throughout this file), please drop the `keys=`.

(Maybe you got it here as an artifact of a bad merge?)

File: datastore/tests/unit/test_client.py [hunk 28]
```diff
@@ -934,6 +934,17 @@ def test_transaction_defaults(self):
             self.assertIs(xact, mock_klass.return_value)
             mock_klass.assert_called_once_with(client)
 
+    def test_read_only_transaction_defaults(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_client.py]: Once you've deleted `Client.read_only_transaction()` and added `**kwargs` to `Client.transaction()` you can convert this test to cover the case where you pass `read_only=True`.

File: datastore/tests/unit/test_client.py [hunk 29]
```diff
@@ -934,6 +934,17 @@ def test_transaction_defaults(self):
...
+        creds = _make_credentials()
+        client = self._make_one(credentials=creds)
+
+        patch = mock.patch(
+            'google.cloud.datastore.client.Transaction', spec=['__call__'])
```
Comment (dhermes) [inline datastore/tests/unit/test_client.py]: The `Transaction` constructor has no network requirements, so why do you mock it?

File: datastore/tests/unit/test_transaction.py [hunk 30]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
         self.assertIsNone(xact.id)
         self.assertEqual(ds_api.begin_transaction.call_count, 1)
 
+    def test_default(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: You don't need to add this test case, `test_ctor_defaults` already exists.

File: datastore/tests/unit/test_transaction.py [hunk 31]
```diff
@@ -22,12 +22,18 @@ class TestTransaction(unittest.TestCase):
     @staticmethod
     def _get_target_class():
         from google.cloud.datastore.transaction import Transaction
-
         return Transaction
 
+    def _get_options_class(self, **kw):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: :+1: This adheres to our "import at run time" style

File: datastore/tests/unit/test_transaction.py [hunk 32]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
 
+    def test_default(self):
+        project = 'PROJECT'
+        id_ = 850302
+        ds_api = _make_datastore_api(xact=id_)
+        client = _Client(project, datastore_api=ds_api)
+        xact = self._make_one(client)
+        self.assertTrue(isinstance(xact.options, self._get_options_class()))
+
+    def test_read_only(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: Name instead `test_constructor_read_only`

File: datastore/tests/unit/test_transaction.py [hunk 33]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        client = _Client(project, datastore_api=ds_api)
+        read_only = self._get_options_class().ReadOnly()
+        options = self._make_options(read_only=read_only)
+        xact = self._make_one(client, read_only=True)
+        xact.begin()
+        begin = ds_api.begin_transaction
+        begin.assert_called_once()
+        self.assertEqual(begin.call_count, 1)
+
+    def test_read_write(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: Delete this test? It doesn't use `options` and isn't supported by the constructor.

File: datastore/tests/unit/test_transaction.py [hunk 34]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+
+    def test_read_only(self):
+        project = 'PROJECT'
+        id_ = 850302
+        ds_api = _make_datastore_api(xact=id_)
+        client = _Client(project, datastore_api=ds_api)
+        read_only = self._get_options_class().ReadOnly()
+        options = self._make_options(read_only=read_only)
+        xact = self._make_one(client, read_only=True)
+        xact.begin()
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: You don't need to test `begin()`

File: datastore/tests/unit/test_transaction.py [hunk 35]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        self.assertTrue(isinstance(xact.options, self._get_options_class()))
+
+    def test_read_only(self):
+        project = 'PROJECT'
+        id_ = 850302
+        ds_api = _make_datastore_api(xact=id_)
+        client = _Client(project, datastore_api=ds_api)
+        read_only = self._get_options_class().ReadOnly()
+        options = self._make_options(read_only=read_only)
+        xact = self._make_one(client, read_only=True)
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: You should check that `xact._options == options`

File: datastore/tests/unit/test_transaction.py [hunk 36]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        client = _Client(project, datastore_api=ds_api)
+        options = self._make_options(read_write=
+                                     self._get_options_class().ReadWrite())
+        xact = self._make_one(client)
+        xact.begin()
+        begin = ds_api.begin_transaction
+        begin.assert_called_once()
+        self.assertEqual(begin.call_count, 1)
+
+    def test_previous_tid(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This test doesn't correspond to any code changes, please remove it.

File: datastore/tests/unit/test_transaction.py [hunk 37]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        client = _Client(project, datastore_api=ds_api)
+        options_klass = self._get_options_class()
+        options = self._make_options(read_only=options_klass.ReadOnly())
+        entity = _Entity()
+        xact = self._make_one(client, read_only=True)
+        xact.begin()
+        with self.assertRaises(RuntimeError):
+            xact.put(entity)
+
+    def test_put_read_write(self):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: This test doesn't correspond to any code changes, please remove it.

File: datastore/tests/unit/test_transaction.py [hunk 38]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        options = self._make_options(read_write=read_write)
+        entity = _Entity()
+        xact = self._make_one(client)
+        xact.begin()
+        xact.put(entity)
+        mutated_entity = _mutated_pb(self, xact.mutations, 'insert')
+        self.assertEqual(mutated_entity.key, entity.key.to_protobuf())
+
+
+def _mutated_pb(test_case, mutation_pb_list, mutation_type):
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: Please remove this helper (only used by `test_put_read_write`, which should be removed)

File: datastore/tests/unit/test_transaction.py [hunk 39]
```diff
@@ -212,6 +218,95 @@ class Foo(Exception):
...
+        begin.assert_called_once()
+        self.assertEqual(begin.call_count, 1)
+
+    def test_put_read_only(self):
+        project = 'PROJECT'
+        id_ = 943243
+        ds_api = _make_datastore_api(xact_id=id_)
+        client = _Client(project, datastore_api=ds_api)
+        options_klass = self._get_options_class()
+        options = self._make_options(read_only=options_klass.ReadOnly())
```
Comment (dhermes) [inline datastore/tests/unit/test_transaction.py]: `options` is unused.

File: datastore/google/cloud/datastore/client.py (hunk 16 above)
Comment (chemelnucfin) [reply datastore/google/cloud/datastore/client.py]: I get errors relating to the tests if I don't change it though.  Should I modify the tests instead?

File: datastore/google/cloud/datastore/transaction.py (hunk 19 above)
Comment (chemelnucfin) [reply datastore/google/cloud/datastore/transaction.py]: I was assuming that read_only is False by default, so I made that the first option.  I can change it if you want.

File: datastore/google/cloud/datastore/client.py (hunk 16 above)
Comment (dhermes) [reply datastore/google/cloud/datastore/client.py]: That's because you've edited the `gapic` generated file. Please revert that edit (see my comment there).

File: datastore/google/cloud/datastore/transaction.py (hunk 19 above)
Comment (dhermes) [reply datastore/google/cloud/datastore/transaction.py]: I want

File: datastore/google/cloud/datastore/client.py (hunk 16 above)
Comment (chemelnucfin) [reply datastore/google/cloud/datastore/client.py]: The errors are this:
https://gist.github.com/chemelnucfin/eee00b41ba2ac8b8f10e1d68bb3047f6
and this:
https://gist.github.com/chemelnucfin/4b010867914d44473b2994b7b8cd6317

Comment (dhermes) [reply datastore/google/cloud/datastore/client.py]: That's because of your edit in the `gapic` generated file.

**Comment by tseaver** [comment]
@chemelnucfin  [Lint failure](https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/4460).

**Comment by chemelnucfin** [comment]
@tseaver PTAL, thanks.

File: datastore/tests/unit/test_client.py [hunk 40]
```diff
@@ -934,6 +934,15 @@ def test_transaction_defaults(self):
...
             mock_klass.assert_called_once_with(client)
 
+    def test_read_only_transaction_defaults(self):
+        from google.cloud.datastore.transaction import Transaction
+        from google.cloud.datastore_v1.types import TransactionOptions        
+        creds = _make_credentials()
+        client = self._make_one(credentials=creds)
+        xact = client.transaction(read_only=True)
+        self.assertIsInstance(xact._options.read_only,
+                              TransactionOptions.ReadOnly)
```
Comment (dhermes) [inline datastore/tests/unit/test_client.py]: Just assert equality here. `__eq__` on protobuf objects will be defined as expected.

File: datastore/tests/unit/test_client.py [hunk 41]
```diff
@@ -940,8 +940,8 @@ def test_read_only_transaction_defaults(self):
         creds = _make_credentials()
         client = self._make_one(credentials=creds)
         xact = client.transaction(read_only=True)
-        self.assertIsInstance(xact._options.read_only,
-                              TransactionOptions.ReadOnly)
+        self.assertEqual(xact._options.read_only,
+                         TransactionOptions.ReadOnly())
```
Comment (dhermes) [inline datastore/tests/unit/test_client.py]: @chemelnucfin Assert equality for all of `xact._options`

Comment (dhermes) [reply datastore/tests/unit/test_client.py]: Your latest commit still isn't doing `self.assertEqual(xact._options, foo)`

File: datastore/google/cloud/datastore/client.py [hunk 42]
```diff
@@ -511,9 +511,13 @@ def batch(self):
         """Proxy to :class:`google.cloud.datastore.batch.Batch`."""
         return Batch(self)
 
-    def transaction(self):
-        """Proxy to :class:`google.cloud.datastore.transaction.Transaction`."""
-        return Transaction(self)
+    def transaction(self, **kwargs):
+        """Proxy to :class:`google.cloud.datastore.transaction.Transaction`.
+
+            :type kwargs: dict
```
Comment (dhermes) [inline datastore/google/cloud/datastore/client.py]: Sorry I just noticed this. It should be dedented to align with the `"""`.

**Comment by chemelnucfin** [comment]
@tseaver any comments?
